### PR TITLE
[REVIEW] add pair_iterator and factories, tests

### PR DESCRIPTION
**Pull request by karthikeyann**
closes https://github.com/rapidsai/cudf/issues/3489

### CODE REVIEW

File: cpp/include/cudf/detail/iterator.cuh [hunk 1]
```diff
@@ -134,6 +207,55 @@ auto make_null_replacement_iterator(column_device_view const& column,
...
+  return thrust::make_transform_iterator(
+      thrust::counting_iterator<cudf::size_type>{0},
+      null_replaced_pair_accessor<Element>{column, null_replacement});
+}
+
+/**
+ * @brief Constructs a pair iterator over a column's values and its validity.
+ *
+ * Dereferencing the returned iterator for element `i` will return
+ * `pair(column[i], true)`. This iterator is only allowed for non-nullable
```
Comment (harrism) [inline cpp/include/cudf/detail/iterator.cuh]: Why isn't there a version for nullable columns that returns the validity of the element rather than `true`?

File: cpp/include/cudf/detail/iterator.cuh [hunk 2]
```diff
@@ -80,6 +80,79 @@ struct null_replaced_value_accessor
   }
 };
```
Comment (harrism) [inline cpp/include/cudf/detail/iterator.cuh]: I notice that you have an accessor that replaces the value of null elements, and one that only works for non-nullable elements. What about an accessor that just returns the value and validity of elements in a nullable column without replacing them? Is this not needed?

Comment (karthikeyann) [reply cpp/include/cudf/detail/iterator.cuh]: Element at null position, is an invalid data. I don't see any use case where they would want read this. 
I will update `make_pair_iterator` for this as well.

File: cpp/include/cudf/detail/iterator.cuh (hunk 1 above)
Comment (karthikeyann) [reply cpp/include/cudf/detail/iterator.cuh]: Same case as other comment. I will update `make_pair_iterator`.

File: cpp/include/cudf/detail/iterator.cuh (hunk 2 above)
Comment (karthikeyann) [reply cpp/include/cudf/detail/iterator.cuh]: added

Comment (harrism) [reply cpp/include/cudf/detail/iterator.cuh]: If it's not needed, why did you add it? I was only asking why so I can understand the usage.

File: cpp/include/cudf/detail/iterator.cuh [hunk 3]
```diff
@@ -237,23 +262,26 @@ auto make_null_replacement_pair_iterator(column_device_view const& column,
...
  * @throws `cudf::logic_error` if column datatype and Element type mismatch.
  *
  * @tparam Element The type of elements in the column
+ * @tparam has_nulls boolean indicating to treat the column is nullable
  * @param column The column to iterate
  * @return auto Iterator that returns valid column elements, and validity of the
  * element in a pair
  */
-template <typename Element>
+template <bool has_nulls, typename Element>
```
Comment (harrism) [inline cpp/include/cudf/detail/iterator.cuh]: Why not put the template arguments in the same order as in the `pair_accessor` struct?

File: cpp/include/cudf/detail/iterator.cuh [hunk 4]
```diff
@@ -134,6 +232,58 @@ auto make_null_replacement_iterator(column_device_view const& column,
...
+{
+  return thrust::make_transform_iterator(
+      thrust::counting_iterator<cudf::size_type>{0},
+      null_replaced_pair_accessor<Element>{column, null_replacement});
+}
+
+/**
+ * @brief Constructs a pair iterator over a column's values and its validity.
+ *
+ * Dereferencing the returned iterator for element `i` will return
```
Comment (jrhemstad) [inline cpp/include/cudf/detail/iterator.cuh]:
```suggestion
 * Dereferencing the returned iterator returns a `thrust::pair<Element, bool>`.
 * 
 * If an element at position `i` is valid (or `has_nulls == false`), then for `p = *(iter + i)`, `p.first` contains 
 * the value of the element at `i` and `p.second == true`.
 *
 * Else, if the element at `i` is null, then the value of `p.first` is undefined and `p.second == false`. 
```

File: cpp/include/cudf/detail/iterator.cuh [hunk 5]
```diff
@@ -134,6 +232,58 @@ auto make_null_replacement_iterator(column_device_view const& column,
...
+ * @throws `cudf::logic_error` if column datatype and Element type mismatch.
+ *
+ * @tparam Element The type of elements in the column
+ * @param column The column to iterate
+ * @param null_replacement The value to return for null elements
+ * @return auto Iterator that returns valid column elements, or a null
+ * replacement value for null elements and validity of the element in a pair
+ */
+template <typename Element>
+auto make_null_replacement_pair_iterator(column_device_view const& column,
```
Comment (jrhemstad) [inline cpp/include/cudf/detail/iterator.cuh]: Is there a use case for a null replaced pair iterator?

File: cpp/include/cudf/detail/iterator.cuh [hunk 6]
```diff
@@ -134,6 +232,58 @@ auto make_null_replacement_iterator(column_device_view const& column,
...
+ * @throws `cudf::logic_error` if column datatype and Element type mismatch.
+ *
+ * @tparam Element The type of elements in the column
+ * @tparam has_nulls boolean indicating to treat the column is nullable
+ * @param column The column to iterate
+ * @return auto Iterator that returns valid column elements, and validity of the
+ * element in a pair
+ */
+template <typename Element, bool has_nulls=false> 
+auto make_pair_iterator(column_device_view const& column)
```
Comment (jrhemstad) [inline cpp/include/cudf/detail/iterator.cuh]: We also need one of these that works for `cudf::scalar`.

File: cpp/tests/utilities/cudf_gmock.hpp [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+#pragma once
+
+//TODO after legacy are removed
+//#include "cudf_gtest.hpp"
+//#include <gmock/gmock.h>
+
+#define ASSERT_CUDA_SUCCEEDED(expr) ASSERT_EQ(cudaSuccess, expr)
+#define EXPECT_CUDA_SUCCEEDED(expr) EXPECT_EQ(cudaSuccess, expr)
+
+#define ASSERT_CUDF_SUCCEEDED(gdf_error_expression)                           \
```
Comment (jrhemstad) [inline cpp/tests/utilities/cudf_gmock.hpp]: These aren't necessary anymore. `gdf_error` codes don't exist.

File: cpp/include/cudf/detail/iterator.cuh (hunk 5 above)
Comment (karthikeyann) [reply cpp/include/cudf/detail/iterator.cuh]: Yes. There is an example use case in unit test.

File: cpp/include/cudf/detail/iterator.cuh (hunk 6 above)
Comment (karthikeyann) [reply cpp/include/cudf/detail/iterator.cuh]: `cudf::scalar` contains a single value. Do we need an _iterator_ for `cudf::scalar`?

Comment (jrhemstad) [reply cpp/include/cudf/detail/iterator.cuh]: Yes, several places have needed this functionality (e.g., https://github.com/rapidsai/cudf/pull/3471). Just make a `thrust::constant_iterator` around the scalar's value.

File: cpp/include/cudf/detail/iterator.cuh (hunk 5 above)
Comment (jrhemstad) [reply cpp/include/cudf/detail/iterator.cuh]: Let me rephrase. Is there an actual _need_ for it?

Comment (karthikeyann) [reply cpp/include/cudf/detail/iterator.cuh]: not exactly. It is needed for the example use case in unit test. But the example use case can be modified to not to use this iterator at all. 
should I remove and update the unit test?

Comment (jrhemstad) [reply cpp/include/cudf/detail/iterator.cuh]: I generally don't like adding functionality unless there is a clear need for it, because every new feature adds to the cost of maintenance and it's a bummer to have to maintain things that aren't used.

Comment (harrism) [reply cpp/include/cudf/detail/iterator.cuh]: To be explicit: please remove it for now @karthikeyann .

**Comment by harrism** [comment]
rerun tests

File: cpp/include/cudf/detail/iterator.cuh (hunk 6 above)
Comment (karthikeyann) [reply cpp/include/cudf/detail/iterator.cuh]: added.

File: cpp/include/cudf/detail/iterator.cuh (hunk 5 above)
Comment (karthikeyann) [reply cpp/include/cudf/detail/iterator.cuh]: removed.

File: cpp/include/cudf/detail/iterator.cuh [hunk 8]
```diff
@@ -80,6 +81,104 @@ struct null_replaced_value_accessor
...
+  /** -------------------------------------------------------------------------*
+   * @brief constructor
+   * @param[in] _col column device view of cudf column
+   * -------------------------------------------------------------------------**/
+  pair_accessor(column_device_view const& _col)
+      : col{_col}
+  {
+    CUDF_EXPECTS(data_type(experimental::type_to_id<Element>()) == col.type(),
+                 "the data type mismatch");
+    CUDF_EXPECTS(_col.nullable(), "Unexpected non-nullable column.");
```
Comment (trevorsm7) [inline cpp/include/cudf/detail/iterator.cuh]:
```suggestion
    if (has_nulls) {
      CUDF_EXPECTS(_col.nullable(), "Unexpected non-nullable column.");
    }
```

File: cpp/include/cudf/detail/iterator.cuh [hunk 9]
```diff
@@ -80,6 +81,104 @@ struct null_replaced_value_accessor
...
+  }
+
+  CUDA_DEVICE_CALLABLE
+  thrust::pair<Element, bool> operator()(cudf::size_type i) const {
+    return {col.element<Element>(i), col.is_valid_nocheck(i)};
```
Comment (trevorsm7) [inline cpp/include/cudf/detail/iterator.cuh]:
```suggestion
    return {col.element<Element>(i), has_nulls ? col.is_valid_nocheck(i) : false};
```

File: cpp/include/cudf/detail/iterator.cuh [hunk 10]
```diff
@@ -80,6 +81,104 @@ struct null_replaced_value_accessor
...
+    CUDF_EXPECTS(data_type(experimental::type_to_id<Element>()) == col.type(),
+                 "the data type mismatch");
+  }
+
+  CUDA_DEVICE_CALLABLE
+  thrust::pair<Element, bool> operator()(cudf::size_type i) const {
+    return {col.element<Element>(i), true};
+  }
+};
+// @brief specialization of `pair_accessor` for nullable column
```
Comment (trevorsm7) [inline cpp/include/cudf/detail/iterator.cuh]: Since the implementation is otherwise the same, why not branch on `has_nulls` instead of having a specialized template for it?

Comment (karthikeyann) [reply cpp/include/cudf/detail/iterator.cuh]: Yes. That's better. merged them both.

**Comment by karthikeyann** [comment]
rerun tests

File: cpp/include/cudf/detail/iterator.cuh [hunk 11]
```diff
@@ -153,6 +263,53 @@ auto inline make_validity_iterator(column_device_view const& column)
...
+ * Else, if the scalar is null, then the value of `p.first` is undefined and `p.second == false`. 
+ * 
+ * @throws `cudf::logic_error` if scalar datatype and Element type mismatch.
+ *
+ * @tparam Element The type of elements in the scalar
+ * @param scalar_value The scalar to iterate
+ * @return auto Iterator that returns scalar, and validity of the scalar in a pair
+ */
+template <typename Element>
+auto inline make_scalar_pair_iterator(scalar const& scalar_value) {
```
Comment (jrhemstad) [inline cpp/include/cudf/detail/iterator.cuh]: This should just be an overload of `make_pair_iterator` in order to allow transparently using one or the other in a template function, e.g.,
```c++
template <typename ColumnOrScalar>
void foo(ColumnOrScalar x){
  auto iter = make_pair_iterator(x);
}
```
```suggestion
auto inline make_pair_iterator(scalar const& scalar_value) {
```

File: cpp/include/cudf/detail/iterator.cuh [hunk 12]
```diff
@@ -80,6 +81,84 @@ struct null_replaced_value_accessor
...
+ * the return value for element `i` will return `pair(column[i], true)`  
+ * if it is valid, or `pair(null_replacement, false)` if it is null.
+ *
+ * @throws `cudf::logic_error` if the column is not nullable.
+ * @throws `cudf::logic_error` if column datatype and Element type mismatch.
+ *
+ * @tparam Element The type of elements in the column
+ * -------------------------------------------------------------------------**/
+template <typename Element>
+struct null_replaced_pair_accessor: public null_replaced_value_accessor<Element>
```
Comment (jrhemstad) [inline cpp/include/cudf/detail/iterator.cuh]: I believe this isn't needed anymore.

**Comment by karthikeyann** [comment]
rerun tests

**Comment by karthikeyann** [comment]
rerun tests

**Comment by harrism** [comment]
@karthikeyann we know the reason for the CI failures. rerun tests won't help until those are fixed.

File: cpp/include/cudf/detail/iterator.cuh [hunk 13]
```diff
@@ -153,6 +228,53 @@ auto inline make_validity_iterator(column_device_view const& column)
...
+ *
+ * Else, if the scalar is null, then the value of `p.first` is undefined and `p.second == false`. 
+ * 
+ * @throws `cudf::logic_error` if scalar datatype and Element type mismatch.
+ *
+ * @tparam Element The type of elements in the scalar
+ * @param scalar_value The scalar to iterate
+ * @return auto Iterator that returns scalar, and validity of the scalar in a pair
+ */
+template <typename Element>
```
Comment (jrhemstad) [inline cpp/include/cudf/detail/iterator.cuh]: While this may seem odd, it would be convenient for `make_pair_iterator` to be uniform between a `scalar` and a `column`. Adding this template param is necessary for that. 
```suggestion
template <typename Element, bool is_null>
```

File: cpp/include/cudf/detail/iterator.cuh [hunk 14]
```diff
@@ -134,6 +178,37 @@ auto make_null_replacement_iterator(column_device_view const& column,
...
+ * @throws `cudf::logic_error` if column datatype and Element type mismatch.
+ *
+ * @tparam Element The type of elements in the column
+ * @tparam has_nulls boolean indicating to treat the column is nullable
+ * @param column The column to iterate
+ * @return auto Iterator that returns valid column elements, and validity of the
+ * element in a pair
+ */
+template <typename Element, bool has_nulls=false> 
+auto make_pair_iterator(column_device_view const& column)
```
Comment (nvdbaranec) [inline cpp/include/cudf/detail/iterator.cuh]: Would it also make sense to have a function that takes a plain column_view?

Comment (karthikeyann) [reply cpp/include/cudf/detail/iterator.cuh]: I thought about it. But `cudf::column_device_view::create()` returns a unique pointer which will be deleted (and frees children as well) when the `make_pair_iterator` returns.

**Comment by karthikeyann** [comment]
Benchmark results...
```
-----------------------------------------------------------------------------------------------------------
Benchmark                                                 Time             CPU   Iterations UserCounters...
-----------------------------------------------------------------------------------------------------------
Iterator/double_cub_raw/1000/manual_time              0.009 ms        0.031 ms        78036 bytes_per_second=878.373M/s
Iterator/double_cub_raw/10000/manual_time             0.019 ms        0.042 ms        36145 bytes_per_second=3.86384G/s
Iterator/double_cub_raw/100000/manual_time            0.020 ms        0.042 ms        34175 bytes_per_second=36.4757G/s
Iterator/double_cub_raw/1000000/manual_time           0.039 ms        0.058 ms        18049 bytes_per_second=192.142G/s
Iterator/double_cub_raw/10000000/manual_time          0.150 ms        0.169 ms         4675 bytes_per_second=497.429G/s
Iterator/double_cub_iter/1000/manual_time             0.010 ms        0.032 ms        71334 bytes_per_second=802.666M/s
Iterator/double_cub_iter/10000/manual_time            0.020 ms        0.042 ms        34387 bytes_per_second=3.71215G/s
Iterator/double_cub_iter/100000/manual_time           0.021 ms        0.043 ms        32749 bytes_per_second=35.2056G/s
Iterator/double_cub_iter/1000000/manual_time          0.040 ms        0.059 ms        17642 bytes_per_second=188.565G/s
Iterator/double_cub_iter/10000000/manual_time         0.151 ms        0.170 ms         4643 bytes_per_second=494.418G/s
Iterator/double_thrust_raw/1000/manual_time           0.432 ms        0.448 ms         1606 bytes_per_second=17.6404M/s
Iterator/double_thrust_raw/10000/manual_time          0.435 ms        0.453 ms         1606 bytes_per_second=175.365M/s
Iterator/double_thrust_raw/100000/manual_time         0.437 ms        0.453 ms         1599 bytes_per_second=1.70489G/s
Iterator/double_thrust_raw/1000000/manual_time        0.447 ms        0.468 ms         1556 bytes_per_second=16.6619G/s
Iterator/double_thrust_raw/10000000/manual_time       0.576 ms        0.597 ms         1139 bytes_per_second=129.326G/s
Iterator/double_thrust_iter/1000/manual_time          0.428 ms        0.448 ms         1635 bytes_per_second=17.8076M/s
Iterator/double_thrust_iter/10000/manual_time         0.432 ms        0.454 ms         1583 bytes_per_second=176.778M/s
Iterator/double_thrust_iter/100000/manual_time        0.439 ms        0.458 ms         1572 bytes_per_second=1.69863G/s
Iterator/double_thrust_iter/1000000/manual_time       0.452 ms        0.471 ms         1525 bytes_per_second=16.475G/s
Iterator/double_thrust_iter/10000000/manual_time      0.582 ms        0.587 ms         1169 bytes_per_second=127.937G/s
```

File: cpp/benchmarks/iterator/iterator_benchmark.cu [hunk 15]
```diff
@@ -0,0 +1,182 @@
...
+
+#include "../fixture/benchmark_fixture.hpp"
+#include "../synchronization/synchronization.hpp"
+
+#include <cudf/detail/iterator.cuh> // include iterator header
+// for reduction tests
+#include <cub/device/device_reduce.cuh>
+#include <cudf/detail/utilities/device_operators.cuh>
+#include <thrust/device_vector.h>
+
```
Comment (harrism) [inline cpp/benchmarks/iterator/iterator_benchmark.cu]: Do these benchmarks test the pair_iterator at all? If not, why are they in this PR? (confused)

File: cpp/include/cudf/detail/iterator.cuh (hunk 13 above)
Comment (karthikeyann) [reply cpp/include/cudf/detail/iterator.cuh]: since column_device_view has `pair_begin` and `pair_end` now, is it okay to remove this suggestion for uniformity and also remove `make_pair_iterator(column_device_view)`?

File: cpp/include/cudf/column/column_device_view.cuh [hunk 16]
```diff
@@ -312,6 +314,49 @@ class alignas(16) column_device_view : public detail::column_device_view_base {
...
+   * Else, if the element at `i` is null, then the value of `p.first` is
+   * undefined and `p.second == false`.
+   *
+   * @throws `cudf::logic_error` if tparam `has_nulls == true` and 
+   * `has_nulls() == false`
+   * @throws `cudf::logic_error` if column datatype and Element type mismatch.
+   *---------------------------------------------------------------------------**/
+  template <typename T, bool has_nulls>
+  const_pair_iterator<T,has_nulls> pair_begin() const {
+    return const_pair_iterator<T, has_nulls>{
```
Comment (harrism) [inline cpp/include/cudf/column/column_device_view.cuh]: Can't you just call `make_pair_iterator`?

Comment (karthikeyann) [reply cpp/include/cudf/column/column_device_view.cuh]: I followed the same way `begin` and `end` is coded. besides, `pair_end` will involve an addition `make_pair_iterator(*this)+size()`.
which is better way and why?

File: cpp/benchmarks/iterator/iterator_benchmark.cu (hunk 15 above)
Comment (karthikeyann) [reply cpp/benchmarks/iterator/iterator_benchmark.cu]: I added pair_iterator benchmarks as well.

File: cpp/include/cudf/column/column_device_view.cuh (hunk 16 above)
Comment (harrism) [reply cpp/include/cudf/column/column_device_view.cuh]: Generally DRY (don't repeat yourself) is a good guideline. If there's a function with an implementation exactly the same as this, call the function rather than duplicating its implementation, right?

Comment (ChuckHastings) [reply cpp/include/cudf/column/column_device_view.cuh]: Or would it be cleaner to have make_pair_iterator call this?  Or does the combination of pair_begin() and pair_end() make make_pair_iterator obsolete?

Comment (karthikeyann) [reply cpp/include/cudf/column/column_device_view.cuh]: Yes. I made make_pair_iterator call pair_begin. It makes make_pair_iterator obsolete.

File: cpp/include/cudf/detail/iterator.cuh (hunk 13 above)
Comment (karthikeyann) [reply cpp/include/cudf/detail/iterator.cuh]: removed.

File: cpp/include/cudf/column/column_device_view.cuh (hunk 16 above)
Comment (karthikeyann) [reply cpp/include/cudf/column/column_device_view.cuh]: removed obsolete make_pair_iterator.

File: cpp/include/cudf/column/column_device_view.cuh [hunk 17]
```diff
@@ -312,6 +314,49 @@ class alignas(16) column_device_view : public detail::column_device_view_base {
...
+  /**---------------------------------------------------------------------------*
+   * @brief Return a pair iterator to the element following the last element of
+   * the column.
+   *
+   * @throws `cudf::logic_error` if tparam `has_nulls == true` and 
+   * `has_nulls() == false`
+   * @throws `cudf::logic_error` if column datatype and Element type mismatch.
+   *---------------------------------------------------------------------------**/
+  template<typename T, bool has_nulls>
+  const_iterator<T> pair_end() const {
```
[the remark this answers is not in the document]
Comment (harrism) [reply cpp/include/cudf/column/column_device_view.cuh]: @ChuckHastings this should have been a "request changes" review, I think.

File: cpp/include/cudf/detail/iterator.cuh (hunk 14 above)
Comment (karthikeyann) [reply cpp/include/cudf/detail/iterator.cuh]: @nvdbaranec

**Comment by harrism** [comment]
@jrhemstad @ChuckHastings ready for re-review so we can get this merged.

File: cpp/include/cudf/detail/iterator.cuh [hunk 18]
```diff
@@ -153,6 +154,53 @@ auto inline make_validity_iterator(column_device_view const& column)
...
+ * Else, if the scalar is null, then the value of `p.first` is undefined and `p.second == false`. 
+ * 
+ * @throws `cudf::logic_error` if scalar datatype and Element type mismatch.
+ *
+ * @tparam Element The type of elements in the scalar
+ * @param scalar_value The scalar to iterate
+ * @return auto Iterator that returns scalar, and validity of the scalar in a pair
+ */
+template <typename Element>
+auto inline make_pair_iterator(scalar const& scalar_value) {
```
Comment (nvdbaranec) [inline cpp/include/cudf/detail/iterator.cuh]: I really need a make_pair_iterator() that takes either a column_view or a column_device_view and also returns the same thing as this scalar version  (thrust::pair<Element, bool>).

The overall idea is to be able to access a column's values and validity vector in a generic way while also being able to treat a scalar as a virtual column,  so that functions like these can operated on in a generic way.

do_some_work(column_view left, column_view right)
do_some_work(column_view left, scalar right)
do some_work(scalar left, column_view right)

Comment (harrism) [reply cpp/include/cudf/detail/iterator.cuh]: This PR has taken ages and it is blocking other PRs merging. Please file a separate issue for this and/or create a PR for what you need.

Comment (karthikeyann) [reply cpp/include/cudf/detail/iterator.cuh]: @nvdbaranec I will bring back `make_pair_iterator(column_device_view)`. For  column_view, please create a separate issue.

**Comment by karthikeyann** [comment]
rerun tests

File: cpp/include/cudf/detail/iterator.cuh (hunk 18 above)
Comment (harrism) [reply cpp/include/cudf/detail/iterator.cuh]: @nvdbaranec do you really need make_pair_iterator(column_device_view), or would you be OK with `column_device_view::pair_begin()` and `pair_end()`? I prefer the latter.

Comment (nvdbaranec) [reply cpp/include/cudf/detail/iterator.cuh]: Preference would be make_pair_iterator.  The context here is the refactor of copy_if_else() / scalars.  Specifically, the cleanest refactoring I can do here: https://github.com/rapidsai/cudf/pull/3471#pullrequestreview-323362862

involves wrapping the various inputs (column/column, scalar/column, column/scalar, scalar/scalar) at the highest possible level with iterators and sending those down to the detail implementation.   There's one templated function between the public implementation and the detail layer that takes generic Left and Right types which can be either scalars or column_views - being able to just pass either type to make_pair_iterator() would be pretty clean.

**Comment by karthikeyann** [comment]
@jrhemstad

File: cpp/include/cudf/detail/iterator.cuh [hunk 19]
```diff
@@ -153,6 +183,54 @@ auto inline make_validity_iterator(column_device_view const& column)
...
+ * Else, if the scalar is null, then the value of `p.first` is undefined and `p.second == false`. 
+ * 
+ * @throws `cudf::logic_error` if scalar datatype and Element type mismatch.
+ *
+ * @tparam Element The type of elements in the scalar
+ * @tparam is_null boolean unused
+ * @param scalar_value The scalar to iterate
+ * @return auto Iterator that returns scalar, and validity of the scalar in a pair
+ */
+template <typename Element, bool is_null=false>
```
Comment (jrhemstad) [inline cpp/include/cudf/detail/iterator.cuh]: The `is_null` value should be checked against `scalar_value.is_valid()`.

Comment (karthikeyann) [reply cpp/include/cudf/detail/iterator.cuh]: `is_null` tparam is added just to keep similar interface as `make_pair_iterator(column_device_view)` as a requirement https://github.com/rapidsai/cudf/pull/3490#discussion_r362013093. It is not supposed to be used.

Comment (karthikeyann) [reply cpp/include/cudf/detail/iterator.cuh]: I removed the name `is_null` to avoid confusion.

Comment (nvdbaranec) [reply cpp/include/cudf/detail/iterator.cuh]: Couple of comments:

- I like the idea of having the parameter actually being named.  It actually serves a real purpose, so the user should know it's there.

- Related, I think it probably shouldn't have a default value, similar to the column_device_view.  The danger there is the user actually has a scalar that's null, but is quietly wrapping as if it weren't.

- I agree with Jake on verifying the template parameter against the actual is_null() of the scalar at construction time. 

- It seems like a good idea to use the template parameter to stuff the validity value in the returned pairs instead of always calling is_null().

Comment (karthikeyann) [reply cpp/include/cudf/detail/iterator.cuh]: @nvdbaranec 
`has_nulls` tparam for `make_pair_iterator(column_device_view)` is present for 2 purposes - 
1. prevent accidentally access invalid null_mask ptr. 
2. run time speed up we could get for iterators by doing compile time optimization for non-nullable columns. 

In `make_pair_iterator(scalar)`, both purposes are not required. `is_valid` always is stored in `scalar` and `is_valid()` value is always returned for pair iterator. The user is not quietly wrapping it because we return a pair(value, validity).

In fact, adding `is_null==is_valid()` check does additional data transfer to host. (`is_null` is different from `has_nulls`).

Comment (nvdbaranec) [reply cpp/include/cudf/detail/iterator.cuh]: Ah, you're right.  This would only make sense if the scalar kept a cpu-side cached version of _is_valid around.  Then the user could cheaply do an if() to create a <true> or <false> pair iterator.  This is kind of like the cached null count in a column_view and calling has_nulls().

**Comment by kkraus14** [comment]
@jrhemstad this is marked ready to review but you have changes requested still. Could you review/approve when you get the chance or dismiss your review?
